### PR TITLE
make CollectionType#find_by_gid work with correct global id format

### PR DESCRIPTION
**Pull request by no-reply**
addresses #4696 by:

  - allowing `CollectionType#find_by_gid` to find "real" GlobalID objects and matching strings.
  - providing a rake task to convert existing references to the new syntax
  - deprecating `#gid` in favor of `#to_global_id`

to move away from the old custom format, we'll need to provide a data migration/rake task that can fix existing references.

@samvera/hyrax-code-reviewers

### CODE REVIEW

File: app/services/hyrax/collections/migration_service.rb
Comment (cjcolvar) [inline app/services/hyrax/collections/migration_service.rb:33]: It looks like the `.to_s` here is unnecessary because it is being called in `#collection_type_gid=`

File: app/services/hyrax/collections/migration_service.rb
Comment (cjcolvar) [inline app/services/hyrax/collections/migration_service.rb:77]: Do you need to use `.to_s` on the result of `.to_global_id` because `.collection_type_gid` is a string and not a `GlobalID`?

File: lib/generators/hyrax/templates/db/seeds.rb
Comment (cjcolvar) [inline lib/generators/hyrax/templates/db/seeds.rb:65]: The `.to_s` appears to be unnecessary because it is called in `collection_type_gid=`.

File: spec/tasks/rake_spec.rb
Comment (cjcolvar) [inline spec/tasks/rake_spec.rb:58]: Nice!

**Review by cjcolvar — COMMENTED** [review]
This looks really good!  I'm happy to fully approve once my question on MigrationService line 77 is resolved.

File: lib/generators/hyrax/templates/db/seeds.rb
Comment (no-reply) [reply lib/generators/hyrax/templates/db/seeds.rb]: thanks. not sure how this snuck in on me

File: app/services/hyrax/collections/migration_service.rb
Comment (no-reply) [reply app/services/hyrax/collections/migration_service.rb]: ty!

File: app/services/hyrax/collections/migration_service.rb
Comment (no-reply) [reply app/services/hyrax/collections/migration_service.rb:77]: yeah, that's right.

File: app/controllers/hyrax/dashboard/collections_controller.rb
Comment (elrayle) [inline app/controllers/hyrax/dashboard/collections_controller.rb:63]: I like the change to #to_global_id.  Makes the purpose of the method clearer.

**Review by elrayle — APPROVED** [review]
This looks really good.  I like seeing it move to use Rails provided Global ID processing.

File: app/services/hyrax/collections/migration_service.rb
Comment (elrayle) [reply app/services/hyrax/collections/migration_service.rb:77]: I gave my approval for this, although it looks like this change is still pending.

**Review by elrayle — CHANGES_REQUESTED** [review]
@cjcolvar request needs to be resolved before merging.

File: app/services/hyrax/collections/migration_service.rb
Comment (elrayle) [reply app/services/hyrax/collections/migration_service.rb:77]: I wonder if there should be a test for this.  Seems like one of the tests should have failed when collection.collection_type_gid didn't equal Hyrax::CollectionType.find_or_create_default_collection_type.to_global_id because the first was a String and the later was a GlobalID.

Comment (elrayle) [reply app/services/hyrax/collections/migration_service.rb:77]: Looking at this more, the line after this sets `collection.collection_type_gid` using `to_global_id` which appears to imply that the gid in the collection is a GlobalID.

Comment (elrayle) [reply app/services/hyrax/collections/migration_service.rb:77]: And the line 63 in the collections_controller also sets it to GlobalID...
```
@collection.collection_type_gid = CollectionType.find(collection_type_id).to_global_id
```

Is this an issue for backward compatibility where collection_type_gid is a String?

**Review by elrayle — APPROVED** [review]
Per conversation with @no-reply in Slack, all comment by @cjcolvar were resolved.  This change looks good to me.

File: app/services/hyrax/collections/migration_service.rb
Comment (no-reply) [reply app/services/hyrax/collections/migration_service.rb:77]: > Is this an issue for backward compatibility where collection_type_gid is a String?

`#collection_type_gid=` casts to a string, right?

Comment (cjcolvar) [reply app/services/hyrax/collections/migration_service.rb:77]: I was concerned about this line because it looks like it is a String != GlobalID comparison.  Am I misreading this line or will a GlobalID automatically cast so it isn't an issue?

Comment (no-reply) [reply app/services/hyrax/collections/migration_service.rb:77]: let's just render the issue moot and revert the change in this line and the test for this service.

the whole thing is dead code that probably should have been deprecated for removal somewhere around Hyrax 2.2.0. https://github.com/samvera/hyrax/pull/4716

Comment (cjcolvar) [reply app/services/hyrax/collections/migration_service.rb:83]: Good point.  Sounds good to me.

**Review by jeremyf — APPROVED** [review]

**Pull request merged**
